pennylane/data/attributes/dictionary.py

```diff
@@ -0,0 +1,92 @@
+# Copyright 2018-2023 Xanadu Quantum Technologies Inc.
+
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+
+#     http://www.apache.org/licenses/LICENSE-2.0
+
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""Contains an AttributeType that allows for heterogenous dictionaries
+of Dataset attributes."""
+
+
+import typing
+from collections.abc import Mapping, MutableMapping
+from typing import Dict, Generic, Optional, TypeVar, Union
+
+from pennylane.data.base.attribute import AttributeType
+from pennylane.data.base.mapper import MapperMixin
+from pennylane.data.base.typing_util import HDF5Any, HDF5Group, T
+
+
+class DatasetDict(
+    Generic[T],
+    AttributeType[HDF5Group, typing.Mapping[str, T], Optional[typing.Mapping[str, T]]],
+    MutableMapping,
+    MapperMixin,
+):
+    """Provides a dict-like collection for Dataset attribute types."""
+
+    Self = TypeVar("Self", bound="DatasetDict")
+
+    type_id = "dict"
+
+    def __post_init__(self, value: Optional[typing.Mapping[str, T]], info):
+        if value:
+            self.update(value)
+
+    def default_value(self) -> None:
+        return None
+
+    def hdf5_to_value(self, bind: HDF5Group) -> typing.MutableMapping[str, T]:
+        return self
+
+    def value_to_hdf5(self, bind_parent: HDF5Group, key: str, value: None) -> HDF5Group:
+        grp = bind_parent.create_group(key)
+
+        return grp
+
+    def copy_value(self) -> Dict[str, T]:
+        return {key: attr.copy_value() for key, attr in self._mapper.items()}
+
+    def copy(self) -> Dict[str, T]:
+        """Returns a copy if this mapping as a builtin ``dict``, with all
+        elements copied."""
+        return self.copy_value()
+
+    def __getitem__(self, __key: str) -> T:
+        return self._mapper[__key].get_value()
+
+    def __setitem__(self, __key: str, __value: Union[T, AttributeType[HDF5Any, T, T]]) -> None:
+        self._mapper[__key] = __value
```

Comment:
Similar to List,
```python
>>> import pennylane as qml
>>> x = qml.data.attributes.DatasetDict({"1": 2, "3": 4})
>>> x["1"] = 5
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/Users/mikhail/ro/pennylane/pennylane/data/attributes/dictionary.py", line 66, in __setitem__
    self._mapper[__key] = __value
  File "/Users/mikhail/ro/pennylane/pennylane/data/base/mapper.py", line 117, in __setitem__
    self.set_item(key, value, None)
  File "/Users/mikhail/ro/pennylane/pennylane/data/base/mapper.py", line 112, in set_item
    attr_type(value, info, parent_and_key=(self.bind, key))
  File "/Users/mikhail/ro/pennylane/pennylane/data/base/attribute.py", line 246, in __init__
    self._bind = self._set_value(value, info, parent, key)
  File "/Users/mikhail/ro/pennylane/pennylane/data/base/attribute.py", line 315, in _set_value
    new_bind = self.value_to_hdf5(parent, key, value)
  File "/Users/mikhail/ro/pennylane/pennylane/data/attributes/scalar.py", line 33, in value_to_hdf5
    bind_parent[key] = value
  File "/Users/mikhail/ro/pennylane/venv/lib/python3.9/site-packages/h5py/_hl/group.py", line 484, in __setitem__
    h5o.link(ds.id, self.id, name, lcpl=lcpl)
  File "h5py/_objects.pyx", line 54, in h5py._objects.with_phil.wrapper
  File "h5py/_objects.pyx", line 55, in h5py._objects.with_phil.wrapper
  File "h5py/h5o.pyx", line 202, in [h5py.h5o.link](http://h5py.h5o.link/)
OSError: Unable to create link (name already exists)
```

Reply:
right, `h5py` is a bit more picky about overwriting existing objects than `zarr`. I'll fix this